### PR TITLE
Add Version#compare

### PR DESCRIPTION
**Pull request by dduugg**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----
As a follow-up to https://github.com/Homebrew/brew/pull/15131#discussion_r1156203167 this PR adds `Version#compare`. It may seem like a minor thing, but IMHO this is much easier for the reader to follow along with (which is also why I'm not in turn calling `public_send` within `compare`). I'm open to alternative method names here, e.g. `compare?` is probably more in line with Ruby idioms.

Since version.rb was already overly long, I've also moved `HeadVersion` to a new file and moved it under the `Version` namespace (consistent with how `NullVersion` is structured). There is only a single reference to `HeadVersion`, which is already within `Version`, so I don't believe other changes are necessary.

cc @carlocab

### CODE REVIEW

File: Library/Homebrew/version/head.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# typed: true
+# frozen_string_literal: true
+
+class Version
+  # A formula's HEAD version.
+  # @see https://docs.brew.sh/Formula-Cookbook#unstable-versions-head Unstable versions (head)
+  #
+  # @api private
+  class HeadVersion < Version
```
Comment (reitermarkus) [inline Library/Homebrew/version/head.rb:9]: Should this be nested like `Version::HeadVersion`? Seems weird.

Comment (reitermarkus) [reply Library/Homebrew/version/head.rb:9]: Also do we actually need this class? Seems like adding the methods here to `Version` would be simple enough.

Comment (apainintheneck) [reply Library/Homebrew/version/head.rb:9]: Seems to already be a pattern with the Version class.

https://github.com/Homebrew/brew/blob/70451ea7b14f1305bc149601e8ed74f7467c1753/Library/Homebrew/version/null.rb#L6-L10

https://github.com/Homebrew/brew/blob/70451ea7b14f1305bc149601e8ed74f7467c1753/Library/Homebrew/version/parser.rb#L4-L6

I also think refactoring that is probably out of the scope of this PR.

**Review by apainintheneck — APPROVED** [review]
LGTM! I remember running into this recently and thought it was kind of ugly.

**Review by MikeMcQuaid — APPROVED** [review]
Looks great thanks again!

**Pull request merged**